packages/rest/src/body-parser.ts

```diff
@@ -0,0 +1,287 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/rest
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {OperationObject} from '@loopback/openapi-v3-types';
+import * as debugModule from 'debug';
+import * as HttpErrors from 'http-errors';
+import {Request, Response, RequestBodyParserOptions} from './types';
+
+import {
+  json,
+  urlencoded,
+  text,
+  OptionsJson,
+  OptionsUrlencoded,
+  OptionsText,
+  Options,
+} from 'body-parser';
+import {inject} from '@loopback/context';
+import {SchemaObject, ReferenceObject, isReferenceObject} from '..';
+import {RestBindings} from './keys';
+import {is} from 'type-is';
+import {RestHttpErrors} from './rest-http-error';
+
+type HttpError = HttpErrors.HttpError;
+
+const debug = debugModule('loopback:rest:body-parser');
+
+export type RequestBody = {
+  // tslint:disable:no-any
+  value: any | undefined;
+  coercionRequired?: boolean;
+  mediaType?: string;
+  schema?: SchemaObject | ReferenceObject;
+};
+
+/**
+ * Get the content-type header value from the request
+ * @param req Http request
+ */
+function getContentType(req: Request): string | undefined {
+  return req.get('content-type');
+}
+
+/**
+ * Express body parser function type
+ */
+type BodyParserWithCallback = (
+  request: Request,
+  response: Response,
+  callback: (err: HttpError) => void,
+) => void;
+
+function normalizeParsingError(err: HttpError) {
+  debug('Cannot parse request body %j', err);
+  if (!err.statusCode || err.statusCode >= 500) {
+    err.statusCode = 400;
+  }
+  return err;
+}
+
+/**
+ * Parse the body asynchronously
+ * @param handle The express middleware handler
+ * @param request Http request
+ */
+function parse(
+  handle: BodyParserWithCallback,
+  request: Request,
+): Promise<void> {
+  // A hack to fool TypeScript as we don't need `response`
+  const response = ({} as any) as Response;
+  return new Promise<void>((resolve, reject) => {
+    handle(request, response, err => {
+      if (err) {
+        reject(normalizeParsingError(err));
+        return;
+      }
+      resolve();
+    });
+  });
+}
+
+// Default limit of the body length
+const DEFAULT_LIMIT = '1mb';
+
+type ParserOption<T extends 'json' | 'urlencoded' | 'text'> = T extends 'json'
+  ? OptionsJson
+  : T extends 'urlencoded'
+    ? OptionsUrlencoded
+    : T extends 'text' ? OptionsText : Options;
+
+function getParserOptions<T extends 'json' | 'urlencoded' | 'text'>(
+  type: T,
+  options: RequestBodyParserOptions,
+): ParserOption<T> {
+  const opts: {[name: string]: any} = {};
+  Object.assign(opts, options[type], options);
+  for (const k of ['json', 'urlencoded', 'text']) {
+    delete opts[k];
+  }
+  return opts as ParserOption<T>;
+}
+
+/**
+ * Interface to be implemented by body parser extensions
+ */
+export interface BodyParser {
```

Comment:
Please move this interface definition to the top of this source file. It's one of the most important public artifacts provided by this file. It's a popular convention to put public artifacts at the top and implementation details below.

Even better, create a new directory for all things related to body parsing and put the interface into a new "types" file used by body parsing, i.e. `packages/rest/src/body-parser/body-parser.types.ts`

Reply:
Fixed